pkg/extprom/http/instrument_server.go

```diff
@@ -90,17 +95,72 @@ func NewInstrumentationMiddleware(reg prometheus.Registerer, buckets []float64)
 // value. http_requests_total is a metric vector partitioned by HTTP method
 // (label name "method") and HTTP status code (label name "code").
 func (ins *defaultInstrumentationMiddleware) NewHandler(handlerName string, handler http.Handler) http.HandlerFunc {
-	return promhttp.InstrumentHandlerDuration(
-		ins.requestDuration.MustCurryWith(prometheus.Labels{"handler": handlerName}),
-		promhttp.InstrumentHandlerRequestSize(
-			ins.requestSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
-			promhttp.InstrumentHandlerCounter(
-				ins.requestsTotal.MustCurryWith(prometheus.Labels{"handler": handlerName}),
-				promhttp.InstrumentHandlerResponseSize(
-					ins.responseSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
-					handler,
-				),
+	return promhttp.InstrumentHandlerRequestSize(
+		ins.requestSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+		promhttp.InstrumentHandlerCounter(
+			ins.requestsTotal.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+			promhttp.InstrumentHandlerResponseSize(
+				ins.responseSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+				http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					now := time.Now()
+
+					wd := &responseWriterDelegator{w: w}
+					handler.ServeHTTP(wd, r)
+
+					observer := ins.requestDuration.WithLabelValues(
+						wd.Status(),
+						handlerName,
+						strings.ToLower(r.Method),
+					)
+					observer.Observe(time.Since(now).Seconds())
+
+					// If we find a tracingID we'll expose it as Exemplar.
+					span := opentracing.SpanFromContext(r.Context())
+					if span != nil {
+						spanCtx, ok := span.Context().(jaeger.SpanContext)
+						if ok {
+							observer.(prometheus.ExemplarObserver).ObserveWithExemplar(
```

Comment:
Like in the weaveworks example https://github.com/weaveworks/common/pull/202/files#diff-29d7cb6219a01fc80b731ac8004ac0f5edd5a1be507b164939ba4a4d1caf9ac2R70, I think we also need to check for sampled trace ID.